[BUG] quicktype-core 24.0.0: published .d.ts files break strict builds (phantom 'command-line-args' type import, untyped urijs/readable-stream)

## Issue Type

Bug in `quicktype-core`'s published TypeScript declarations

## Context (Environment, Version, Language)

CLI, npm, or app.quicktype.io: npm package `quicktype-core`
Version: 24.0.0
Node: 20.19.4 / 22.23.1 / 24.6.0
TypeScript: 5.8.3

## Description

Compiling a TypeScript project that depends on `quicktype-core` with `skipLibCheck: false` fails on errors inside the package's own published `.d.ts` files:

1. **Phantom type-only dependency**: `dist/RendererOptions/types.d.ts` contains

   ```ts
   import type { OptionDefinition as CommandLineArgsOptionDefinition } from "command-line-args";
   ```

   but `command-line-args` is a dependency of the `quicktype` CLI package, **not** of `quicktype-core`, so the import cannot resolve in a consumer's `node_modules`:

   ```
   node_modules/quicktype-core/dist/RendererOptions/types.d.ts(2,74): error TS2307: Cannot find module 'command-line-args' or its corresponding type declarations.
   ```

2. **Untyped dependencies referenced from public declarations**: `dist/input/JSONSchemaInput.d.ts`, `dist/input/io/NodeIO.d.ts`, and `dist/input/io/get-stream/index.d.ts` import from `urijs` and `readable-stream`, which ship no type declarations, producing implicit-`any` errors (TS7016) under strict mode.

## Expected Behaviour / Output

`tsc` with `"strict": true, "skipLibCheck": false` compiles a project using the documented `quicktype-core` API without errors.

## Current Behaviour / Output

```
node_modules/quicktype-core/dist/RendererOptions/types.d.ts(2,74): error TS2307: Cannot find module 'command-line-args' or its corresponding type declarations.
node_modules/quicktype-core/dist/input/JSONSchemaInput.d.ts(1,17): error TS7016: Could not find a declaration file for module 'urijs'. ...
node_modules/quicktype-core/dist/input/io/NodeIO.d.ts(1,26): error TS7016: Could not find a declaration file for module 'readable-stream'. ...
node_modules/quicktype-core/dist/input/io/get-stream/index.d.ts(1,31): error TS7016: Could not find a declaration file for module 'readable-stream'. ...
```

## Steps to Reproduce

```bash
mkdir repro && cd repro
npm init -y
npm install quicktype-core@24.0.0 typescript@5.8 @types/node
cat > main.ts <<'EOF'
import { quicktype, InputData, jsonInputForTargetLanguage } from "quicktype-core";
async function main() {
    const jsonInput = jsonInputForTargetLanguage("typescript");
    await jsonInput.addSource({ name: "Person", samples: ['{"name": "Alice"}'] });
    const inputData = new InputData();
    inputData.addInput(jsonInput);
    console.log((await quicktype({ inputData, lang: "typescript" })).lines.join("\n"));
}
main();
EOF
npx tsc --strict --module commonjs --target es2020 --esModuleInterop main.ts
```

Workarounds that both fix it: set `skipLibCheck: true`, or `npm i -D @types/command-line-args @types/urijs @types/readable-stream`. Neither should be required.

## Possible Solution

- For (1): don't re-export `command-line-args` types from `quicktype-core` — inline the `OptionDefinition` shape or move the CLI-facing option types into the CLI package. Alternatively add `@types/command-line-args` as a regular dependency of `quicktype-core`.
- For (2): add `@types/urijs` and `@types/readable-stream` as dependencies (not devDependencies) since the public `.d.ts` surface references them, or avoid exposing those types in the public declarations.

Runtime behavior is unaffected — this is purely a typechecking problem. Found while testing the 24.0.0 release on Node 20/22/24.

--- a/package.json
+++ b/package.json
@@ -32,6 +32,7 @@
     ],
     "dependencies": {
         "@glideapps/ts-necessities": "^2.2.3",
+        "@types/readable-stream": "4.0.10",
         "chalk": "^4.1.2",
         "collection-utils": "^1.0.1",
         "command-line-args": "^5.2.1",

--- a/packages/quicktype-core/package.json
+++ b/packages/quicktype-core/package.json
@@ -15,6 +15,8 @@
     },
     "dependencies": {
         "@glideapps/ts-necessities": "2.2.3",
+        "@types/readable-stream": "4.0.10",
+        "@types/urijs": "^1.19.25",
         "browser-or-node": "^3.0.0",
         "collection-utils": "^1.0.1",
         "is-url": "^1.2.4",
@@ -34,11 +36,8 @@
         "@types/node": "~20.19.0",
         "@types/pako": "^1.0.0",
         "@types/pluralize": "0.0.30",
-        "@types/readable-stream": "4.0.10",
         "@types/unicode-properties": "^1.3.0",
-        "@types/urijs": "^1.19.25",
         "@types/wordwrap": "^1.0.3",
-        "command-line-args": "^5.2.1",
         "typescript": "~5.8.3"
     },
     "files": ["dist"],

--- a/packages/quicktype-core/src/RendererOptions/types.ts
+++ b/packages/quicktype-core/src/RendererOptions/types.ts
@@ -12,10 +12,11 @@ export type OptionType = "string" | "boolean" | "enum";
 // that package is a dependency of the quicktype CLI, not of quicktype-core,
 // so the type-only import leaked into the published declarations and broke
 // consumers that compile with `skipLibCheck: false` (issue #2904).  The
-// fields below that have no quicktype-specific meaning (`alias`, `group`,
-// `defaultOption`, `lazyMultiple`, `type`) mirror the previously inherited
-// fields of command-line-args' `OptionDefinition`, so option definitions
-// remain directly consumable by `commandLineArgs` in the CLI.
+// `alias` and `defaultOption` fields below mirror the previously inherited
+// command-line-args fields of the same names, so the CLI's option
+// definitions remain directly consumable by `commandLineArgs` (the CLI adds
+// `type` itself at the call site).  The other previously inherited fields
+// (`group`, `lazyMultiple`, `type`) were dropped as unused.
 export interface OptionDefinition<Name extends string = string, T = unknown> {
     /** Option Name */
     name: Name;
@@ -23,12 +24,6 @@ export interface OptionDefinition<Name extends string = string, T = unknown> {
     alias?: string;
     /** Whether the CLI treats this option as the default positional argument */
     defaultOption?: boolean;
-    /** CLI option group(s) this option belongs to */
-    group?: string | string[];
-    /** Whether repeated CLI values require repeating the option name */
-    lazyMultiple?: boolean;
-    /** Converts a CLI string input to the option's value type */
-    type?: (input: string) => unknown;
     /** Option Description */
     description: string;
     /** Category of Option */

--- a/packages/quicktype-core/src/RendererOptions/types.ts
+++ b/packages/quicktype-core/src/RendererOptions/types.ts
@@ -1,5 +1,4 @@
 import type { EnumOption, Option } from "./index";
-import type { OptionDefinition as CommandLineArgsOptionDefinition } from "command-line-args";
 
 /**
  * Primary options show up in the web UI in the "Language" settings tab,
@@ -9,10 +8,27 @@ import type { OptionDefinition as CommandLineArgsOptionDefinition } from "comman
 export type OptionKind = "primary" | "secondary" | "cli";
 export type OptionType = "string" | "boolean" | "enum";
 
-export interface OptionDefinition<Name extends string = string, T = unknown>
-    extends CommandLineArgsOptionDefinition {
+// This interface used to extend command-line-args' `OptionDefinition`, but
+// that package is a dependency of the quicktype CLI, not of quicktype-core,
+// so the type-only import leaked into the published declarations and broke
+// consumers that compile with `skipLibCheck: false` (issue #2904).  The
+// fields below that have no quicktype-specific meaning (`alias`, `group`,
+// `defaultOption`, `lazyMultiple`, `type`) mirror the previously inherited
+// fields of command-line-args' `OptionDefinition`, so option definitions
+// remain directly consumable by `commandLineArgs` in the CLI.
+export interface OptionDefinition<Name extends string = string, T = unknown> {
     /** Option Name */
     name: Name;
+    /** Single-character CLI alias, e.g. `-o` for `--out` */
+    alias?: string;
+    /** Whether the CLI treats this option as the default positional argument */
+    defaultOption?: boolean;
+    /** CLI option group(s) this option belongs to */
+    group?: string | string[];
+    /** Whether repeated CLI values require repeating the option name */
+    lazyMultiple?: boolean;
+    /** Converts a CLI string input to the option's value type */
+    type?: (input: string) => unknown;
     /** Option Description */
     description: string;
     /** Category of Option */

--- a/test/unit/core-package.test.ts
+++ b/test/unit/core-package.test.ts
@@ -39,6 +39,69 @@ function findNodeImports(directory: string): string[] {
     return offenders;
 }
 
+// Matches module specifiers that can appear in declaration files: static
+// imports/re-exports (`from "urijs"`), inline type imports (`import("urijs")`),
+// and triple-slash type references (`/// <reference types="node" />`).
+const declarationSpecifierPattern =
+    /(?:from\s+|import\s*\(\s*)["']([^"']+)["']|\/\/\/\s*<reference\s+types\s*=\s*["']([^"']+)["']/g;
+
+function findDeclarationFiles(directory: string): string[] {
+    const files: string[] = [];
+
+    for (const entry of fs.readdirSync(directory, { withFileTypes: true })) {
+        const fullPath = path.join(directory, entry.name);
+        if (entry.isDirectory()) {
+            files.push(...findDeclarationFiles(fullPath));
+        } else if (entry.isFile() && entry.name.endsWith(".d.ts")) {
+            files.push(fullPath);
+        }
+    }
+
+    return files;
+}
+
+// "lodash/fp" -> "lodash", "@glideapps/ts-necessities/dist/x" -> "@glideapps/ts-necessities"
+function packageNameOfSpecifier(specifier: string): string {
+    const parts = specifier.split("/");
+    return specifier.startsWith("@") ? parts.slice(0, 2).join("/") : parts[0];
+}
+
+// "@glideapps/ts-necessities" -> "@types/glideapps__ts-necessities"
+function typesPackageFor(packageName: string): string {
+    return packageName.startsWith("@")
+        ? `@types/${packageName.slice(1).replace("/", "__")}`
+        : `@types/${packageName}`;
+}
+
+function packageShipsTypes(
+    packageName: string,
+    fromDirectory: string,
+): boolean {
+    let manifestPath: string;
+    try {
+        manifestPath = require.resolve(`${packageName}/package.json`, {
+            paths: [fromDirectory],
+        });
+    } catch {
+        return false;
+    }
+
+    const manifest = JSON.parse(fs.readFileSync(manifestPath, "utf8")) as {
+        exports?: unknown;
+        types?: string;
+        typings?: string;
+    };
+    if (manifest.types !== undefined || manifest.typings !== undefined) {
+        return true;
+    }
+
+    if (JSON.stringify(manifest.exports ?? null).includes('"types"')) {
+        return true;
+    }
+
+    return fs.existsSync(path.join(path.dirname(manifestPath), "index.d.ts"));
+}
+
 describe("quicktype-core package", () => {
     // Importing the built quicktype-core must not print anything: CLI users
     // redirect stdout (`quicktype ... > out.ts`), so any stray output
@@ -67,4 +130,71 @@ describe("quicktype-core package", () => {
     test("does not use node:-prefixed imports in browser-compatible source", () => {
         expect(findNodeImports(coreSourceDirectory)).toEqual([]);
     });
+
+    // Every bare module specifier in the published .d.ts files must resolve
+    // to type declarations out of the package's own runtime dependencies,
+    // because consumers who compile with `skipLibCheck: false` typecheck
+    // those files.  quicktype-core 24.0.0 shipped a type-only import of
+    // "command-line-args" (a dependency of the CLI, not of quicktype-core;
+    // TS2307 for consumers) and imports of the untyped "urijs" and
+    // "readable-stream" whose @types packages were only devDependencies
+    // (TS7016 under strict) — see
+    // https://github.com/glideapps/quicktype/issues/2904.
+    test.each([
+        ".",
+        "packages/quicktype-core",
+        "packages/quicktype-graphql-input",
+        "packages/quicktype-typescript-input",
+    ])(
+        "declaration files in %s only reference dependencies that resolve with types",
+        (relativePackageDirectory) => {
+            const packageDirectory = path.join(
+                repositoryRoot,
+                relativePackageDirectory,
+            );
+            const declarationFiles = findDeclarationFiles(
+                path.join(packageDirectory, "dist"),
+            );
+            expect(declarationFiles.length).toBeGreaterThan(0);
+
+            const manifest = JSON.parse(
+                fs.readFileSync(
+                    path.join(packageDirectory, "package.json"),
+                    "utf8",
+                ),
+            ) as { dependencies: Record<string, string> };
+            const dependencies = manifest.dependencies;
+
+            const offenders: string[] = [];
+            for (const declarationFile of declarationFiles) {
+                const contents = fs.readFileSync(declarationFile, "utf8");
+                for (const match of contents.matchAll(
+                    declarationSpecifierPattern,
+                )) {
+                    const specifier = match[1] ?? match[2];
+                    if (specifier.startsWith(".")) continue;
+
+                    const packageName = packageNameOfSpecifier(specifier);
+                    const problem =
+                        dependencies[packageName] === undefined
+                            ? "is not a runtime dependency"
+                            : dependencies[typesPackageFor(packageName)] ===
+                                    undefined &&
+                                !packageShipsTypes(
+                                    packageName,
+                                    packageDirectory,
+                                )
+                              ? "has no type declarations"
+                              : undefined;
+                    if (problem !== undefined) {
+                        offenders.push(
+                            `${path.relative(repositoryRoot, declarationFile)}: "${specifier}" ${problem}`,
+                        );
+                    }
+                }
+            }
+
+            expect(offenders).toEqual([]);
+        },
+    );
 });

